src/scratch.js

```diff
@@ -0,0 +1,36 @@
+const DB = require('./db')
+const { asyncFirstSeries } = require('./util/async')
+
+const ENCODING_OPTS = { keyEncoding: 'binary', valueEncoding: 'binary' }
+
+/**
+ * An in-memory wrap over `DB` with a backend DB
+ * which will be queried when a key is not found
+ * in the in-memory scratch. This class is used to implement
+ * checkpointing functionality in CheckpointTrie.
+ */
+module.exports = class ScratchDB extends DB {
+  constructor (db) {
+    super()
+    this._backend = db._db
+  }
+
+  /**
+   * Similar to `DB.get`, but first searches in-memory
+   * scratch DB, if key not found, searches backend DB.
+   */
+  get (key, cb) {
+    const getDBs = this._backend ? [this._db, this._backend] : [this._db]
+    const dbGet = (db, cb2) => {
+      db.get(key, ENCODING_OPTS, (err, v) => {
+        if (err || !v) {
+          cb2(null, null)
+        } else {
+          cb2(null, v)
+        }
+      })
+    }
+
+    asyncFirstSeries(getDBs, dbGet, cb)
+  }
+}
```

Comment:
Had to do some longer look to understand how this is working but getting it now, had some trouble to grasp which DB is used on which occasion! 😄

Looks good and really elegant way to achieve. Does this have any performance implications? Did you measure?

Reply:
Yeah it's still not as clear... Maybe we can iterate on this further.

No I didn't, but intuitively this shouldn't affect performance by much, as the approach is really similar to what was there before, and the changes are mostly structural.